ksqldb-engine/src/test/java/io/confluent/ksql/query/PullQueryWriteStreamTest.java

```diff
@@ -0,0 +1,166 @@
+
+  private static final QueryId TEST_ID = new QueryId("test");
+  private static final int TEST_TIMEOUT_SECONDS = 60;
+  private static final int LIMIT = 5;
+
+  @Rule
+  public final Timeout timeout = Timeout.seconds(TEST_TIMEOUT_SECONDS);
+
+  @Mock private LogicalSchema schema;
+  private PullQueryWriteStream writeStream;
+  private ScheduledExecutorService executorService;
+
+  @Before
+  public void setUp() {
+    final StreamedRowTranslator translator = new StreamedRowTranslator(schema, Optional.empty());
+    writeStream = new PullQueryWriteStream(OptionalInt.of(LIMIT), translator);
+    writeStream.write(ImmutableList.of(StreamedRow.header(TEST_ID, schema)));
+
+    executorService = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  @After
+  public void tearDown() {
+    executorService.shutdown();
+  }
+
+  @Test
+  public void shouldAcceptWrites() {
+    // When:
+    writeStream.write(getData(1));
+
+    // Then:
+    assertThat(writeStream.getTotalRowsQueued(), is(1));
+    assertThat(writeStream.size(), is(1));
+  }
+
+  @Test
+  public void shouldPoll() {
+    // Given:
+    final AtomicBoolean written = new AtomicBoolean(false);
+    writeStream.write(getData(1), ignored -> written.set(true));
+
+    // When:
+    final KeyValueMetadata<List<?>, GenericRow> row = writeStream.poll();
+
+    // Then:
+    assertThat(row.getKeyValue().value().get(0), is(0));
+    assertThat("expected written callback to be called", written.get());
+    assertThat("expected writeStream to be empty after poll", writeStream.isEmpty());
+  }
+
+  @Test
+  public void shouldHandleLimit() {
+    // Given:
+    final AtomicBoolean limit = new AtomicBoolean(false);
+    final AtomicBoolean end = new AtomicBoolean(false);
+
+    writeStream.setLimitHandler(() -> limit.set(true));
+    writeStream.setCompletionHandler(() -> end.set(true));
+
+    // When:
+    writeStream.write(getData(LIMIT + 1));
+
+    // Then:
+    assertThat("expected limit handler to be called", limit.get());
+    assertThat("expected end handler to be called", end.get());
+    assertThat("expected queue to be done", writeStream.isDone());
+    assertThat(writeStream.size(), is(LIMIT)); // should not enqueue more than LIMIT
+  }
+
+  @Test
+  public void shouldCallDrainHandlerWhenHasCapacity() {
+    // Given:
+    final AtomicBoolean called = new AtomicBoolean(false);
+
+    writeStream.setWriteQueueMaxSize(1);
+    writeStream.write(getData(1));
+    writeStream.drainHandler(ignored -> called.set(true));
+
+    // When:
+    writeStream.poll();
+
+    // Then:
+    assertThat("expected drain handler to be called", called.get());
+  }
+
+  @Test
+  public void shouldAwaitCapacity() throws InterruptedException {
+    // Given:
+    writeStream.setWriteQueueMaxSize(1);
+    writeStream.write(getData(1));
+
+    // When:
+    executorService.schedule(() -> writeStream.poll(), 500, TimeUnit.MILLISECONDS);
+    writeStream.awaitCapacity(TEST_TIMEOUT_SECONDS * 2, TimeUnit.SECONDS);
```

Comment:
Maybe unify 500 ms / TEST_TIMEOUT_SECONDS seconds somehow?  Meh...

Reply:
Also, can we use lower numbers;)?

Reply:
I like to keep the test timeout high because when debugging if I'm stepping through things I want to have at least a minute before JUnit barfs on me. The expected path is pretty quick